File: compiler/frontend/src/org/jetbrains/kotlin/psi/KtClassInitializer.java
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.psi;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.kotlin.lexer.KtTokens;
import org.jetbrains.kotlin.psi.stubs.KotlinPlaceHolderStub;
import org.jetbrains.kotlin.psi.stubs.elements.KtStubElementTypes;

public class KtClassInitializer extends KtDeclarationStub<KotlinPlaceHolderStub<KtClassInitializer>> implements KtStatementExpression {
    public KtClassInitializer(@NotNull ASTNode node) {
        super(node);
    }

    public KtClassInitializer(@NotNull KotlinPlaceHolderStub<KtClassInitializer> stub) {
        super(stub, KtStubElementTypes.ANONYMOUS_INITIALIZER);
    }

    @Override
    public <R, D> R accept(@NotNull KtVisitor<R, D> visitor, D data) {
        return visitor.visitAnonymousInitializer(this, data);
    }

    @Nullable
    public KtExpression getBody() {
        return findChildByClass(KtExpression.class);
    }

    @Nullable
    public PsiElement getOpenBraceNode() {
        KtExpression body = getBody();
        return (body instanceof KtBlockExpression) ? ((KtBlockExpression) body).getLBrace() : null;
    }

    @NotNull
    public PsiElement getOpenBraceNodeOrSelf() {
        PsiElement result = getOpenBraceNode();
        return result != null ? result : this;
    }

    @Nullable
    public PsiElement getInitKeyword() {
        return findChildByType(KtTokens.INIT_KEYWORD);
    }
}

File: compiler/psi/psi-api/src/org/jetbrains/kotlin/psi/KtClassInitializer.kt
/*
 * Copyright 2010-2026 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */

package org.jetbrains.kotlin.psi

import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import org.jetbrains.kotlin.KtStubBasedElementTypes
import org.jetbrains.kotlin.lexer.KtTokens
import org.jetbrains.kotlin.psi.psiUtil.getParentOfType
import org.jetbrains.kotlin.psi.stubs.KotlinPlaceHolderStub
import org.jetbrains.kotlin.utils.sure

/**
 * Represents an init block in a class that executes during instance initialization.
 *
 * ### Example:
 *
 * ```kotlin
 * class Foo {
 *     init {
 *         println("initialized")
 *     }
 * //  ^________________________^
 * //  The entire 'init' block
 * }
 * ```
 */
@OptIn(KtImplementationDetail::class)
class KtClassInitializer : KtDeclarationStub<KotlinPlaceHolderStub<KtClassInitializer>>, KtAnonymousInitializer {
    @KtImplementationDetail
    constructor(node: ASTNode) : super(node)

    @KtImplementationDetail
    constructor(stub: KotlinPlaceHolderStub<KtClassInitializer>) : super(stub, KtStubBasedElementTypes.CLASS_INITIALIZER)

    override fun <R, D> accept(visitor: KtVisitor<R, D>, data: D): R = visitor.visitClassInitializer(this, data)

    override val body: KtExpression?
        get() = findChildByClass(KtExpression::class.java)

    /**
     * The opening brace of the `init` block's body, or `null` if the body is not a block.
     */
    val openBraceNode: PsiElement?
        get() = (body as? KtBlockExpression)?.lBrace

    /**
     * The `init` keyword introducing this block.
     */
    val initKeyword: PsiElement
        get() = findChildByType(KtTokens.INIT_KEYWORD)!!

    /** The [KtClassOrObject] that declares this `init` block. */
    override val containingDeclaration: KtClassOrObject
        get() = getParentOfType<KtClassOrObject>(true).sure { "Should only be present in class or object" }
}
